File: trunk/symmetric-core/src/main/java/org/jumpmind/symmetric/file/FileSyncZipDataWriter.java
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.jumpmind.symmetric.file;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import org.apache.commons.io.IOUtils;
import org.apache.commons.lang.StringEscapeUtils;
import org.apache.commons.lang.StringUtils;
import org.jumpmind.db.model.Table;
import org.jumpmind.exception.IoException;
import org.jumpmind.symmetric.io.data.Batch;
import org.jumpmind.symmetric.io.data.CsvData;
import org.jumpmind.symmetric.io.data.DataContext;
import org.jumpmind.symmetric.io.data.DataEventType;
import org.jumpmind.symmetric.io.data.IDataWriter;
import org.jumpmind.symmetric.io.stage.IStagedResource;
import org.jumpmind.symmetric.model.FileSnapshot;
import org.jumpmind.symmetric.model.FileSnapshot.LastEventType;
import org.jumpmind.symmetric.model.FileTrigger;
import org.jumpmind.symmetric.model.FileTriggerRouter;
import org.jumpmind.symmetric.model.Node;
import org.jumpmind.symmetric.service.IFileSyncService;
import org.jumpmind.symmetric.service.INodeService;
import org.jumpmind.util.FormatUtils;
import org.jumpmind.util.Statistics;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class FileSyncZipDataWriter implements IDataWriter {

    static final Logger log = LoggerFactory.getLogger(FileSyncZipDataWriter.class);

    protected long byteCount;
    protected long maxBytesToSync;
    protected IFileSyncService fileSyncService;
    protected IStagedResource stagedResource;
    protected ZipOutputStream zos;
    protected Table snapshotTable;
    protected Batch batch;
    protected Map<Batch, Statistics> statistics = new HashMap<Batch, Statistics>();
    protected List<FileSnapshot> snapshotEvents;
    protected DataContext context;
    protected INodeService nodeService;

    public FileSyncZipDataWriter(long maxBytesToSync, IFileSyncService fileSyncService,
            INodeService nodeService, IStagedResource stagedResource) {
        this.maxBytesToSync = maxBytesToSync;
        this.fileSyncService = fileSyncService;
        this.stagedResource = stagedResource;
        this.nodeService = nodeService;
    }

    public void open(DataContext context) {
        this.context = context;
    }

    public void close() {
    }

    public Map<Batch, Statistics> getStatistics() {
        return statistics;
    }

    public void start(Batch batch) {
        this.batch = batch;
        this.statistics.put(batch, new Statistics());
        this.snapshotEvents = new ArrayList<FileSnapshot>();
    }

    public boolean start(Table table) {
        this.snapshotTable = table;
        return true;
    }

    public void write(CsvData data) {
        DataEventType eventType = data.getDataEventType();
        if (eventType == DataEventType.INSERT || eventType == DataEventType.UPDATE) {
            Map<String, String> columnData = data.toColumnNameValuePairs(
                    snapshotTable.getColumnNames(), CsvData.ROW_DATA);
            FileSnapshot snapshot = new FileSnapshot();
            snapshot.setTriggerId(columnData.get("TRIGGER_ID"));
            snapshot.setRouterId(columnData.get("ROUTER_ID"));
            snapshot.setFileModifiedTime(FormatUtils.parseDate(
                    columnData.get("FILE_MODIFIED_TIME"), FormatUtils.TIMESTAMP_PATTERNS));
            snapshot.setCrc32Checksum(Long.parseLong(columnData.get("CRC32_CHECKSUM")));
            snapshot.setFileSize(Long.parseLong(columnData.get("FILE_SIZE")));
            snapshot.setLastUpdateBy(columnData.get("LAST_UPDATE_BY"));
            snapshot.setFileName(columnData.get("FILE_NAME"));
            snapshot.setFilePath(columnData.get("FILE_PATH"));
            snapshot.setLastEventType(LastEventType.fromCode(columnData.get("LAST_EVENT_TYPE")));
            snapshotEvents.add(snapshot);
        } else if (eventType == DataEventType.RELOAD) {
            String targetNodeId = context.getBatch().getTargetNodeId();
            Node targetNode = nodeService.findNode(targetNodeId);
            List<FileTriggerRouter> fileTriggerRouters = fileSyncService
                    .getFileTriggerRoutersForCurrentNode();
            for (FileTriggerRouter fileTriggerRouter : fileTriggerRouters) {
                if (fileTriggerRouter.isEnabled()
                        && fileTriggerRouter.isInitialLoadEnabled()
                        && fileTriggerRouter.getRouter().getNodeGroupLink().getTargetNodeGroupId()
                                .equals(targetNode.getNodeGroupId())) {
                    DirectorySnapshot directorySnapshot = fileSyncService
                            .getDirectorySnapshot(fileTriggerRouter);
                    snapshotEvents.addAll(directorySnapshot);
                }
            }
        }
    }

    public void end(Table table) {
    }

    public void end(Batch batch, boolean inError) {

        try {
            if (!inError) {
                if (zos == null) {
                    zos = new ZipOutputStream(stagedResource.getOutputStream());
                }

                Set<String> commands = new HashSet<String>();
                StringBuilder script = new StringBuilder("fileList = new HashMap();\n");
                for (FileSnapshot snapshot : snapshotEvents) {
                    FileTriggerRouter triggerRouter = fileSyncService.getFileTriggerRouter(
                            snapshot.getTriggerId(), snapshot.getRouterId());
                    if (triggerRouter != null) {
                        StringBuilder command = new StringBuilder("\n");
                        LastEventType eventType = snapshot.getLastEventType();

                        FileTrigger fileTrigger = triggerRouter.getFileTrigger();

                        String targetBaseDir = triggerRouter.getTargetBaseDir();
                        if (StringUtils.isBlank(targetBaseDir)) {
                            targetBaseDir = fileTrigger.getBaseDir();
                        }
                        targetBaseDir = StringEscapeUtils.escapeJava(targetBaseDir);

                        command.append("targetBaseDir = \"").append(targetBaseDir).append("\";\n");
                        command.append("processFile = true;\n");
                        command.append("sourceFileName = \"").append(snapshot.getFileName())
                                .append("\";\n");
                        command.append("sourceFilePath = \"");
                        command.append(StringEscapeUtils.escapeJava(snapshot.getFilePath()))
                                .append("\";\n");

                        StringBuilder entryName = new StringBuilder(Long.toString(batch
                                .getBatchId()));
                        entryName.append("/");
                        if (!snapshot.getFilePath().equals(".")) {
                            entryName.append(snapshot.getFilePath()).append("/");
                        }
                        entryName.append(snapshot.getFileName());

                        File file = fileTrigger.createSourceFile(snapshot);
                        if (file.isDirectory()) {
                            entryName.append("/");
                        }

                        if (StringUtils.isNotBlank(fileTrigger.getBeforeCopyScript())) {
                            command.append(fileTrigger.getBeforeCopyScript()).append("\n");
                        }

                        command.append("if (processFile) {\n");

                        switch (eventType) {
                            case CREATE:
                            case MODIFY:
                            case SEED:
                                if (file.exists()) {
                                    command.append("  mv (batchDir + \"/\"");
                                    if (!snapshot.getFilePath().equals(".")) {
                                        command.append(" + sourceFilePath + \"/\"");
                                    }
                                    command.append(" + sourceFileName");
                                    command.append(", ");
                                    StringBuilder targetFile = new StringBuilder(
                                            "targetBaseDir + \"/");
                                    if (!snapshot.getFilePath().equals(".")) {
                                        targetFile.append(StringEscapeUtils.escapeJava(snapshot
                                                .getFilePath()));
                                        targetFile.append("/");
                                    }
                                    targetFile.append(snapshot.getFileName());
                                    targetFile.append("\"");
                                    command.append(targetFile);
                                    command.append(");\n");
                                    command.append("  fileList.put(").append(targetFile)
                                            .append(",\"");
                                    command.append(eventType.getCode());
                                    command.append("\");\n");
                                }
                                break;
                            case DELETE:
                                command.append("  org.apache.commons.io.FileUtils.deleteQuietly(new java.io.File(");
                                StringBuilder targetFile = new StringBuilder("targetBaseDir + \"/");
                                if (!snapshot.getFilePath().equals(".")) {
                                    targetFile.append(StringEscapeUtils.escapeJava(snapshot
                                            .getFilePath()));
                                    targetFile.append("/");
                                }
                                targetFile.append(snapshot.getFileName());
                                targetFile.append("\"");
                                command.append(targetFile);
                                command.append("));\n");
                                command.append("  fileList.put(").append(targetFile).append(",\"");
                                command.append(eventType.getCode());
                                command.append("\");\n");
                                break;
                            default:
                                break;
                        }

                        if (StringUtils.isNotBlank(fileTrigger.getAfterCopyScript())) {
                            command.append(fileTrigger.getAfterCopyScript()).append("\n");
                        }

                        if (!commands.contains(command.toString())) {
                            if (file.exists()) {
                                byteCount += file.length();
                                ZipEntry entry = new ZipEntry(entryName.toString());
                                entry.setSize(file.length());
                                entry.setTime(file.lastModified());
                                zos.putNextEntry(entry);
                                if (file.isFile()) {
                                    FileInputStream fis = new FileInputStream(file);
                                    try {
                                        IOUtils.copy(fis, zos);
                                    } finally {
                                        IOUtils.closeQuietly(fis);
                                    }
                                }
                                zos.closeEntry();
                            } else if (eventType != LastEventType.DELETE) {
                                log.warn(
                                        "Could not find the {} file to package for synchronization.  Skipping it.",
                                        file.getAbsolutePath());
                            }

                            command.append("}\n\n");
                            script.append(command.toString());
                            commands.add(command.toString());

                        }

                    } else {
                        log.error(
                                "Could not locate the file trigger ({}) router ({}) to process a snapshot event.  The event will be ignored",
                                snapshot.getTriggerId(), snapshot.getRouterId());
                    }
                }

                script.append("return fileList;\n");
                ZipEntry entry = new ZipEntry(batch.getBatchId() + "/sync.bsh");
                zos.putNextEntry(entry);
                IOUtils.write(script.toString(), zos);
                zos.closeEntry();

            }
        } catch (IOException e) {
            throw new IoException(e);
        }

    }

    public void finish() {
        try {
            if (zos != null) {
                zos.finish();
                IOUtils.closeQuietly(zos);
            }
        } catch (IOException e) {
            throw new IoException(e);
        } finally {
            stagedResource.close();
            stagedResource.setState(IStagedResource.State.READY);
        }
    }

    public boolean readyToSend() {
        return byteCount > maxBytesToSync;
    }

}

File: symmetric-client-clib/src/file/FileSyncZipDataWriter.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "file/FileSyncZipDataWriter.h"
#include "service/FileSyncService.h"


void SymFileSyncZipDataWriter_open(SymFileSyncZipDataWriter *this) {
    char * ZIP_NAME = "tmp/staging/filesync_outgoing/filesync.zip";

    SymFileUtils_mkdir("tmp/staging/filesync_outgoing/");

    int rc;
    this->zip = zip_open(ZIP_NAME, ZIP_CREATE|ZIP_TRUNCATE, &rc);
    if (!this->zip) {
        zip_error_t error;
        zip_error_init_with_code(&error, rc);
        SymLog_error("Could not start zip %s %s", ZIP_NAME, zip_error_strerror(&error));
        zip_error_fini(&error);
    }
}


static void SymFileSyncZipDataWriter_startBatch(SymFileSyncZipDataWriter *this, SymBatch *batch) {
    if (this->isFirstBatch) {
        this->isFirstBatch = 0;
    }

    this->batch = batch;
    this->snapshotEvents = SymList_new(NULL);
}

static void SymFileSyncZipDataWriter_endBatch(SymFileSyncZipDataWriter *this, SymBatch *batch) {
    SymMap * /*<char*, char*>*/ entries = SymMap_new(NULL, 4);
    SymStringBuilder *script = SymStringBuilder_newWithString("fileList = new HashMap();\n");
    int i;
    for (i = 0; i < this->snapshotEvents->size; ++i) {
        SymFileSnapshot *snapshot = this->snapshotEvents->get(this->snapshotEvents, i);
        SymFileTriggerRouter *triggerRouter =
                this->fileSyncService->getFileTriggerRouter(this->fileSyncService, snapshot->triggerId, snapshot->routerId);
        if (triggerRouter != NULL) {
            SymStringBuilder *command = SymStringBuilder_new("\n");
            char eventType = snapshot->lastEventType[0];

            SymFileTrigger *fileTrigger = triggerRouter->fileTrigger;
            char* targetBaseDir = triggerRouter->targetBaseDir;
            if (SymStringUtils_isBlank(targetBaseDir)) {
                targetBaseDir = fileTrigger->baseDir;
            }

            command->append(command, "targetBaseDir = \"")->append(command, targetBaseDir)->append(command, "\";\n");

            command->append(command, "if (targetBaseDir.startsWith(\"${androidBaseDir}\")) {                      \n");
            command->append(command, "    targetBaseDir = targetBaseDir.replace(\"${androidBaseDir}\", \"\");     \n");
            command->append(command, "    targetBaseDir = androidBaseDir + targetBaseDir;                         \n");
            command->append(command, "} else if (targetBaseDir.startsWith(\"${androidAppFilesDir}\")) {           \n");
            command->append(command, "    targetBaseDir = targetBaseDir.replace(\"${androidAppFilesDir}\", \"\"); \n");
            command->append(command, "    targetBaseDir = androidAppFilesDir + targetBaseDir;                     \n");
            command->append(command, "}                                                                           \n");

            command->append(command, "processFile = true;\n");
            command->append(command, "sourceFileName = \"")->append(command, snapshot->fileName)->append(command, "\";\n");
            command->append(command, "targetRelativeDir = \"");

            if (!SymStringUtils_equals(snapshot->relativeDir, ".")) {
                command->append(command, snapshot->relativeDir);
                command->append(command, "\";\n");
            } else {
                command->append(command, "\";\n");
            }

            command->append(command, "targetFileName = sourceFileName;\n");
            command->append(command, "sourceFilePath = \"");
            command->append(command, snapshot->relativeDir)->append(command, "\";\n");


            SymStringBuilder *entryName = SymStringBuilder_newWithString(SymStringUtils_format("%ld", batch->batchId));
            entryName->append(entryName, "/");
            if (!SymStringUtils_equals(snapshot->relativeDir, ".")) {
                entryName->append(entryName, snapshot->relativeDir)->append(entryName, "/");
            }
            entryName->append(entryName, snapshot->fileName);

            char* path = triggerRouter->fileTrigger->getPath(triggerRouter->fileTrigger, snapshot);
            if (SymFileUtils_isDir(path)) {
                entryName->append(entryName, "/");
            }

            if (SymStringUtils_isNotBlank(fileTrigger->beforeCopyScript)) {
                command->append(command, fileTrigger->beforeCopyScript);
                command->append(command, "\n");
            }
            command->append(command, "if (processFile) {\n");
            char* targetFile = "targetBaseDir + \"/\" + targetRelativeDir + \"/\" + targetFileName";

            switch (eventType) {
            case 'C':
            case 'M':
                if (SymFileUtils_exists(path)) {
                    command->append(command, "  File targetBaseDirFile = new File(targetBaseDir);\n");
                    command->append(command, "  if (!targetBaseDirFile.exists()) {\n");
                    command->append(command, "    targetBaseDirFile.mkdirs();\n");
                    command->append(command, "  }\n");
                    command->append(command, "  java.io.File sourceFile = new java.io.File(batchDir + \"/\"");

                    if (!SymStringUtils_equals(snapshot->relativeDir, ".")) {
                        command->append(command, " + sourceFilePath + \"/\"");
                    }
                    command->append(command, " + sourceFileName");
                    command->append(command, ");\n");

                    command->append(command, "  java.io.File targetFile = new java.io.File(");
                    command->append(command, targetFile);
                    command->append(command, ");\n");

                    // no need to copy directory if it already exists
                    command->append(command, "  if (targetFile.exists() && targetFile.isDirectory()) {\n");
                    command->append(command, "      processFile = false;\n");
                    command->append(command, "  }\n");

                    //                        // conflict resolution
                    //                        FileConflictStrategy conflictStrategy = triggerRouter.getConflictStrategy();
                    //                        if (conflictStrategy == FileConflictStrategy.TARGET_WINS ||
                    //                                conflictStrategy == FileConflictStrategy.MANUAL) {
                    //                            command->append(command, "  if (targetFile.exists() && !targetFile.isDirectory()) {\n");
                    //                            command->append(command, "    long targetChecksum = org.apache.commons.io.FileUtils.checksumCRC32(targetFile);\n");
                    //                            command->append(command, "    if (targetChecksum != " + snapshot.getOldCrc32Checksum() + "L) {\n");
                    //                            if (conflictStrategy == FileConflictStrategy.MANUAL) {
                    //                                command->append(command, "      throw new org.jumpmind.symmetric.file.FileConflictException(targetFileName + \" was in conflict \");\n");
                    //                            } else {
                    //                                command->append(command, "      processFile = false;\n");
                    //                            }
                    //                            command->append(command, "    }\n");
                    //                            command->append(command, "  }\n");
                    //                        }

                    command->append(command, "  if (processFile) {\n");
                    command->append(command, "    if (sourceFile.isDirectory()) {\n");
                    command->append(command, "      org.apache.commons.io.FileUtils.copyDirectory(sourceFile, targetFile, true);\n");
                    command->append(command, "    } else {\n");
                    command->append(command, "      org.apache.commons.io.FileUtils.copyFile(sourceFile, targetFile, true);\n");
                    command->append(command, "    }\n");
                    command->append(command, "  }\n");
                    command->append(command, "  fileList.put(")->append(command, targetFile)->append(command, ",\"");
                    command->appendf(command, "%c", eventType);
                    command->append(command, "\");\n");
                }
                break;
            case 'D':
                command->append(command, "  org.apache.commons.io.FileUtils.deleteQuietly(new java.io.File(");
                command->append(command, targetFile);
                command->append(command, "));\n");
                command->append(command, "  fileList.put(")->append(command, targetFile)->append(command, ",\"");
                command->appendf(command, "%c", eventType);
                command->append(command, "\");\n");
                break;
            default:
                break;
            }

            if (SymStringUtils_isNotBlank(fileTrigger->afterCopyScript)) {
                command->append(command, fileTrigger->afterCopyScript)->append(command, "\n");
            }

            char* previousEventForEntryString = (char*)entries->get(entries, entryName->toString(entryName));

            unsigned short process = 1;
            if (previousEventForEntryString != NULL) {
                char previousEventForEntry = previousEventForEntryString[0];
                if ((previousEventForEntry == eventType)
                        || (previousEventForEntry == 'C' && eventType == 'M')) {
                    process = 0;
                }
            }

            if (process) {
                if (eventType != 'D') {
                    if (SymFileUtils_exists(path)) {
                        long fileSize = SymFileUtils_getFileSize(path);
                        this->byteCount += fileSize;

                        long START_POSITION = 0, END_POSITION = 0;
                        if (!this->zip) {
                            SymLog_error("zip is not initialized, seg fault imminent.");
                        }
                        if (SymFileUtils_isDir(path)) {
                            zip_add_dir(this->zip, entryName->toString(entryName));
                        } else {
                            zip_source_t *zipSource = zip_source_file(this->zip, path, START_POSITION, END_POSITION);
                            if (!zipSource) {
                                SymLog_error("error creating zip source from path '%s': %s\n", path, zip_strerror(this->zip));
                            }
                            long zipIndex = zip_file_add(this->zip, entryName->toString(entryName), zipSource, ZIP_FL_OVERWRITE);
                            if (zipIndex < 0) {
                                SymLog_error("error adding file '%s': %s\n", path, zip_strerror(this->zip));
                            }
                        }
                        entries->put(entries, entryName->toString(entryName), SymStringUtils_format("%c", eventType));
                    } else {
                        SymLog_warn("Could not find the %s file to package for synchronization.  Skipping it.", path);
                    }
                }
                command->append(command, "}\n\n");
                script->append(script, command->destroyAndReturn(command));
                entryName->destroy(entryName);
            }
        } else {
            SymLog_error("Could not locate the file trigger (%s) router (%s) to process a snapshot event.  The event will be ignored",
                    snapshot->triggerId, snapshot->routerId);
        }

    }
    script->append(script, "return fileList;\n");

    {
        char * bshFileName = SymStringUtils_format("%ld/sync.bsh", batch->batchId);
        zip_source_t *zipSourceBuffer = zip_source_buffer(this->zip, script->toString(script), script->size-1, 0);
        if (!zipSourceBuffer) {
            printf("error creating zip source buffer: %s\n", zip_strerror(this->zip));
        }
        long zipIndex = zip_file_add(this->zip, bshFileName, zipSourceBuffer, ZIP_FL_OVERWRITE);
        if (zipIndex < 0) {
            printf("error adding buffer: %s\n", zip_strerror(this->zip));
        }
        free(bshFileName);
    }
    {
        char * batchInfoFileName = SymStringUtils_format("%ld/batch-info.txt", batch->batchId);
        char * channelIdClone = SymStringUtils_format("%s", batch->channelId); // The batch will be freed before the zip gets written.
        zip_source_t *zipSourceBuffer = zip_source_buffer(this->zip, channelIdClone, strlen(channelIdClone), 0);
        if (!zipSourceBuffer) {
            printf("error creating zip source buffer: %s\n", zip_strerror(this->zip));
        }
        long zipIndex = zip_file_add(this->zip, batchInfoFileName, zipSourceBuffer, ZIP_FL_OVERWRITE);
        if (zipIndex < 0) {
            printf("error adding buffer: %s\n", zip_strerror(this->zip));
        }
        free(batchInfoFileName);
//        free(channelIdClone);
    }

//    script->destroy(script);
    entries->destroy(entries);
    this->batch = NULL;
}

static void SymFileSyncZipDataWriter_startTable(SymFileSyncZipDataWriter *this, SymTable *table) {
    this->snapshotTable = table;
}

static void SymFileSyncZipDataWriter_endTable(SymFileSyncZipDataWriter *this, SymTable *table) {
    // no-op.
}

unsigned short SymFileSyncZipDataWriter_write(SymFileSyncZipDataWriter *this, SymCsvData *data) {
    SymDataEventType eventType = data->dataEventType;

    if (eventType == SYM_DATA_EVENT_INSERT || eventType == SYM_DATA_EVENT_UPDATE) {
        //  SymLog_info("data = %s", data->rowData->toString(data->rowData));

        //     SymList * /*<SymColumn>*/ columns = this->snapshotTable->columns;
        SymStringArray *columnNames = this->snapshotTable->getColumnNames(this->snapshotTable);
        // SymLog_info("columnNames = %s", columnNames->toString(columnNames));
        SymMap * /*<String,String>*/ columnData = data->toColumnNameValuePairsRowData(data, columnNames);
        SymMap * /*<String,String>*/ oldColumnData = data->toColumnNameValuePairsOldData(data, columnNames);

        SymFileSnapshot *snapshot = SymFileSnapshot_new(NULL);
        snapshot->triggerId = columnData->get(columnData, "trigger_id");
        snapshot->routerId = columnData->get(columnData, "router_id");
        char* modifedTime = columnData->get(columnData, "file_modified_time");
        if (!SymStringUtils_isBlank(modifedTime)) {
            snapshot->fileModifiedTime = atol(modifedTime);
        }
        char *crc32Checksum = columnData->get(columnData, "crc32_checksum");
        if (!SymStringUtils_isBlank(crc32Checksum)) {
            snapshot->crc32Checksum = atol(crc32Checksum);
        }

        char *oldChecksum = oldColumnData->get(oldColumnData, "crc32_checksum");
        if (!SymStringUtils_isBlank(oldChecksum)) {
            snapshot->oldCrc32Checksum = atol(oldChecksum);
        }
        char *fileSize = columnData->get(columnData, "file_size");
        if (!SymStringUtils_isBlank(fileSize)) {
            snapshot->fileSize = atol(fileSize);
        }
        snapshot->lastUpdateBy = columnData->get(columnData, "last_update_by");
        snapshot->fileName = columnData->get(columnData, "file_name");
        snapshot->relativeDir = columnData->get(columnData, "relative_dir");
        snapshot->lastEventType = columnData->get(columnData, "last_event_type");
        this->snapshotEvents->add(this->snapshotEvents, snapshot);
    } else if (eventType == SYM_DATA_EVENT_RELOAD) {
        char *targetNodeId = this->context->batch->targetNodeId;
        SymNode *targetNode = this->nodeService->findNode(this->nodeService, targetNodeId);
        SymList * /*<SymFileTriggerRouter>*/ fileTriggerRouters =
                this->fileSyncService->getFileTriggerRoutersForCurrentNode(this->fileSyncService);
        int i;
        for (i = 0; i < fileTriggerRouters->size; ++i) {
            SymFileTriggerRouter *fileTriggerRouter = fileTriggerRouters->get(fileTriggerRouters, i);
            if (fileTriggerRouter->enabled
                    && SymStringUtils_equals(fileTriggerRouter->router->nodeGroupLink->targetNodeGroupId,
                            targetNode->nodeGroupId)) {
                SymDirectorySnapshot *directorySnapshot =
                        this->fileSyncService->getDirectorySnapshot(this->fileSyncService, fileTriggerRouter);
                this->snapshotEvents->addAll(this->snapshotEvents, directorySnapshot->fileSnapshots);
            }
        }
    }

    return 0;
}

SymList * SymFileSyncZipDataWriter_getBatchesProcessed(SymFileSyncZipDataWriter *this) {
    return this->super.batchesProcessed;
}

void SymFileSyncZipDataWriter_close(SymFileSyncZipDataWriter *this) {
    int rc = zip_close(this->zip);
    if (rc != 0) {
        SymLog_error("error closing zip file: %s\n", zip_strerror(this->zip));
    }
}

void SymFileSyncZipDataWriter_destroy(SymFileSyncZipDataWriter *this) {
    this->processedTables->destroy(this->processedTables);
    free(this);
}

SymFileSyncZipDataWriter * SymFileSyncZipDataWriter_new(SymFileSyncZipDataWriter *this, char *sourceNodeId) {
    if (this == NULL) {
        this = (SymFileSyncZipDataWriter *) calloc(1, sizeof(SymFileSyncZipDataWriter));
    }
    this->sourceNodeId = sourceNodeId;
    this->isFirstBatch = 1;
    this->processedTables = SymMap_new(NULL, 100);

    SymDataWriter *super = &this->super;

    super->batchesProcessed = SymList_new(NULL);
    super->open = (void *) &SymFileSyncZipDataWriter_open;
    super->startBatch = (void *) &SymFileSyncZipDataWriter_startBatch;
    super->startTable = (void *) &SymFileSyncZipDataWriter_startTable;
    super->write = (void *) &SymFileSyncZipDataWriter_write;
    super->endTable = (void *) &SymFileSyncZipDataWriter_endTable;
    super->endBatch = (void *) &SymFileSyncZipDataWriter_endBatch;
    super->close = (void *) &SymFileSyncZipDataWriter_close;
    super->destroy = (void *) &SymFileSyncZipDataWriter_destroy;
    return this;
}
